### DISCUSSION TITLE
Contol usb barcode scanner beep

### DISCUSSION THREAD

**Discussion by samshipengs** [pyusb/pyusb#447] [Q&A] [answered]
I got the bar code scanner working with pyusb, however sometimes the read value is not complete, for example a bar code128 represents 123456789012, sometimes got read partially such as: 1234, 012 etc. I imagine this might be due to both the lighting conditions, scanning angles, and the quality of the scanner. 

I would like the bar code scanner only beeps after a python script checks the scanned value is valid (e.g. 12 digits long), is there anyway to interact with the scanner in such way using pyusb?

Thanks!

**Comment by mcuee** [comment]
I am not so sure if I understand you correctly or not. But I guess that is not a problem which pyusb can help you. Rather you need to work with the Bar Code Scanner vendor, so that with a certain condition (handshake between your application and the scanner) then it will beep.

Or if you know the communication protocol and assume that the scanner has an API to beep, you can call that API when your python script confirms the scanned value to be valid. 

Or you can install an Arduino or things like that and create a beep device which your python script can control.

**Reply by samshipengs** [reply]
Thanks for the reply and apologies for the vague question. But basically, I started trying to use Pyusb as I need to grab barcode value through barcode scanner into a python prgoram, and I started from this code example: https://github.com/vpatron/barcode_scanner_python/blob/master/usb_scanner_hid_read_demo.py#L166 

although I noticed the `data` the `ep.read(1000, 500)` outputs sometimes is partial, for example a barcode `1234567890` sometimes gets read to `123`. Sometimes there is a follow-up (I guess a buffer of some sort?) in the next `ep.read` call, tha is - `123` first, then `4567890\n` comes in (`\n` is the end of scan chars used by the barcode scanner), but sometimes there is no follow-up, and simply just `123`. 

So in the question I posted above, I was trying to somehow control the usb barcode scanner, such that it only beeps when the scanned code has the right length e.g. 12 (and it won't beep when it's `123` for example).

**Reply by samshipengs** [reply]
I bought a single barcode scanner simply online to just play around, so there is not really working with the vendor I suppose. I can scan some QR code on the instruction manual (came with the purchase) for disabling certain type of barcode or turn on/off, lower/higher beep volume, but not sure how that or conditional beep can be made using python code.

**Reply by mcuee** [reply, marked as answer]
You have to know the communication protocol before you can really use libusb/pyusb effectively. You can try to ask your vendor for documentation/API. If you can not get that then you have to carry out reverse engineering.

Example using pyusb: https://slomkowski.eu/tutorials/eavesdropping-usb-and-writing-driver-in-python/
